File: engine/src/core/game_state/tracking.rs
use super::GameState;

impl GameState {
    pub fn reset_keyword_tracking(&mut self) {
        self.turn1_abilities_played.clear();
        self.turn2_abilities_played.clear();
        self.player1_cheer_blade_heart_count = 0;
        self.player2_cheer_blade_heart_count = 0;
        self.player1.debut_count_this_turn = 0;
        self.player2.debut_count_this_turn = 0;
        self.player1.deck_refreshed_this_turn = false;
        self.player2.deck_refreshed_this_turn = false;
        self.player1.last_resolution_cards.clear();
        self.player2.last_resolution_cards.clear();
        self.clear_card_appearance_tracking();
        self.clear_auto_ability_trigger_tracking();
        self.reset_change_flags();
        self.cheer_check_completed = false;
        self.reset_loop_detection();
        self.player1.deployed_this_turn.clear();
        self.player2.deployed_this_turn.clear();
        self.baton_touch_count_p1 = 0;
        self.baton_touch_count_p2 = 0;
        self.baton_touch_arriving_card_ids.clear();
        self.baton_touch_zero_cost = false;
        self.baton_touch_replaced_member_cost = None;
        self.baton_touch_replaced_member_id = None;
        self.baton_touch_arriving_card_id = None;
        self.clear_area_placement_tracking();
    }

    /// Register a modify_yell_count effect: (player_slot 1|2, delta).
    /// Modifiers are data; the required count is derived on read.
    pub fn add_yell_count_modifier(&mut self, player_slot: u8, delta: i32) {
        self.yell_count_modifiers.push((player_slot, delta));
    }

    /// Derived cheer checks required for `player_id`:
    /// max(0, base + Σ deltas registered for that player).
    pub fn effective_cheer_checks_required(&self, player_id: &str, base: u8) -> u8 {
        let slot = if player_id == self.player1.id || player_id == "p1" {
            1u8
        } else {
            2u8
        };
        let base = self.cheer_check_base.unwrap_or(base);
        let sum: i32 = self
            .yell_count_modifiers
            .iter()
            .filter(|m| m.0 == slot)
            .map(|m| m.1)
            .sum();
        (base as i32 + sum).max(0) as u8
    }

    pub fn perform_cheer_check(&mut self, player_id: &str, blade_count: u8) -> Result<(), String> {
        // The required count is always DERIVED: base (the live's blade count,
        // fixed at the first check) plus every modify_yell_count modifier for
        // this player, whenever it was applied. There is no initialization
        // order dependency: a LiveStart modifier registered before the base
        // exists is simply included in the sum.
        if self.cheer_check_base.is_none() {
            self.cheer_check_base = Some(blade_count);
        }
        self.cheer_checks_required = self.effective_cheer_checks_required(player_id, blade_count);

        let player = if player_id == self.player1.id {
            &mut self.player1
        } else {
            &mut self.player2
        };

        // Q104 / Rule 10.2.1: refresh from waitroom when deck runs out mid-draw.
        let from_bottom = player.yell_from_bottom;
        for _ in 0..blade_count {
            if player.main_deck.cards.is_empty() && !player.waitroom.cards.is_empty() {
                player.refresh();
            }
            // G8: 恋になりたいAQUARIUM makes the yell reveal from the deck bottom.
            let card_id = if from_bottom {
                player.main_deck.draw_bottom()
            } else {
                player.main_deck.draw()
            };
            if let Some(card_id) = card_id {
                self.resolution_zone.cards.push(card_id);
                self.cheer_checks_done += 1;
            }
        }

        if self.cheer_checks_done >= self.cheer_checks_required {
            self.cheer_check_completed = true;
        }
        Ok(())
    }

    pub fn check_required_hearts(&self) -> Result<bool, String> {
        if self.cheer_checks_done < self.cheer_checks_required {
            return Err(format!(
                "Cannot check required hearts: {} of {} cheer checks completed",
                self.cheer_checks_done, self.cheer_checks_required
            ));
        }
        Ok(true)
    }

    pub fn is_action_prohibited(&self, action: &str) -> bool {
        self.prohibition_effects.iter().any(|e| e.contains(action))
    }

}

File: engine_c/src/core/tracking.c
#include "rabuka.h"
#include <string.h>
#include <stdio.h>
#include <stdint.h>

/* Helpers ported from modifiers.rs / abilities.rs — minimal stubs that mirror
   the Rust clear/reset helpers so tracking.rs can call them. */

static void rb_clear_auto_ability_trigger_tracking(GameState *g){
    g->n_auto_ability_trigger_counts = 0;
}

static void rb_reset_change_flags(GameState *g){
    g->position_change_occurred_this_turn = 0;
    g->formation_change_occurred_this_turn = 0;
    g->opponent_live_success_this_turn = 0;
}

static void rb_reset_loop_detection(GameState *g){
    g->n_game_state_history = 0;
    g->loop_detected = 0;
}

static void rb_clear_area_placement_tracking(GameState *g){
    g->n_areas_placed_this_turn = 0;
}

void rb_reset_keyword_tracking(GameState *g){
    g->n_turn1_abilities_played = 0;
    g->n_turn2_abilities_played = 0;
    g->player1_cheer_blade_heart_count = 0;
    g->player2_cheer_blade_heart_count = 0;
    g->p[0].deck_refreshed_this_turn = 0;
    g->p[1].deck_refreshed_this_turn = 0;
    g->last_resolution_cards_p1.n = 0;
    g->last_resolution_cards_p2.n = 0;
    rb_clear_auto_ability_trigger_tracking(g);
    rb_reset_change_flags(g);
    g->cheer_check_completed = 0;
    rb_reset_loop_detection(g);
    g->baton_touch_count_p1 = 0;
    g->baton_touch_count_p2 = 0;
    g->n_baton_touch_arriving_card_ids = 0;
    g->baton_touch_zero_cost = 0;
    g->baton_touch_replaced_member_cost = -1;
    g->baton_touch_replaced_member_id = -1;
    g->baton_touch_arriving_card_id = -1;
    rb_clear_area_placement_tracking(g);
}

void rb_add_yell_count_modifier(GameState *g, uint8_t player_slot, int32_t delta){
    if(g->n_yell_count_modifiers >= 32) return;
    g->yell_count_modifiers[g->n_yell_count_modifiers].slot = player_slot;
    g->yell_count_modifiers[g->n_yell_count_modifiers].delta = delta;
    g->n_yell_count_modifiers++;
}

uint8_t rb_effective_cheer_checks_required(const GameState *g, const char *player_id, uint8_t base){
    uint8_t slot = 2;
    if(player_id && (!strcmp(player_id, "p1") || !strcmp(player_id, "1"))) slot = 1;
    int eff_base = (g->cheer_check_base >= 0) ? g->cheer_check_base : (int)base;
    int sum = 0;
    for(int i=0;i<g->n_yell_count_modifiers;i++){
        if(g->yell_count_modifiers[i].slot == slot) sum += g->yell_count_modifiers[i].delta;
    }
    int total = eff_base + sum;
    if(total < 0) total = 0;
    if(total > 255) total = 255;
    return (uint8_t)total;
}

int rb_perform_cheer_check(GameState *g, const char *player_id, uint8_t blade_count){
    if(g->cheer_check_base < 0){
        g->cheer_check_base = blade_count;
    }
    g->cheer_checks_required = rb_effective_cheer_checks_required(g, player_id, blade_count);

    int pl = 0;
    if(player_id && !strcmp(player_id, "p2")) pl = 1;
    /* mirror Rust: pick player by id; we use p1/p2 string. */
    RbPlayer *player = &g->p[pl];
    int from_bottom = 0; /* TODO: player.yell_from_bottom not ported; assume top */

    for(int i=0;i<blade_count;i++){
        if(player->deck.n==0 && player->discard.n>0){
            /* refresh from waitroom when deck runs out mid-draw — rule 10.2.1 */
            rb_player_refresh(g, pl);
        }
        int card_id = -1;
        if(from_bottom){
            /* draw_bottom not yet ported; fallback to top */
            if(player->deck.n>0) card_id = player->deck.cards[--player->deck.n];
        } else {
            if(player->deck.n>0) card_id = player->deck.cards[--player->deck.n];
        }
        if(card_id != -1){
            if(g->resolution.n < RB_MAX_ZONE) g->resolution.cards[g->resolution.n++] = card_id;
            g->cheer_checks_done++;
        }
    }
    if(g->cheer_checks_done >= g->cheer_checks_required){
        g->cheer_check_completed = 1;
    }
    return 0;
}

int rb_check_required_hearts(const GameState *g){
    if(g->cheer_checks_done < g->cheer_checks_required){
        return -1;
    }
    return 0;
}

int rb_is_action_prohibited(const GameState *g, const char *action){
    if(!action) return 0;
    for(int i=0;i<g->n_prohibition;i++){
        if(strstr(g->prohibition[i], action)) return 1;
    }
    return 0;
}
